tokio/src/sync/notify.rs

```diff
@@ -516,19 +526,44 @@ impl Notify {
             return;
         }
 
-        // At this point, it is guaranteed that the state will not
-        // concurrently change, as holding the lock is required to
-        // transition **out** of `WAITING`.
+        // Move all waiters to a list owned by this function call to ensure atomicity.
+        // This list will still be protected by the `waiters` lock.
+        let waiters_to_notify = UnsafeCell::new(std::mem::take(&mut *waiters));
+        pin!(waiters_to_notify);
```

Comment:
The `LinkedList` type is `Unpin`, so pinning it is a no-op.

Reply:
Yeah, it felt awkward to use Pin here...
My intention was to only shadow the original variable binding to prevent from accidentally moving it. I am not sure was it the best way to express that this variable won't change its address. Anyway, I replaced the pin macro with manual shadowing.